_posts/2024-01-12-kasten-openshift-monitoring.markdown

```diff
@@ -0,0 +1,815 @@
+```
+
+You should get an output like this one 
+```
+{
+  "status": "success",
+  "data": {
+    "resultType": "vector",
+    "result": [
+      {
+        "metric": {
+          "__name__": "up",
+          "container": "kube-rbac-proxy",
+          "endpoint": "metrics",
+          "instance": "10.128.0.38:9154",
+          "job": "dns-default",
+          "namespace": "openshift-dns",
+          "pod": "dns-default-j6n8h",
+          "prometheus": "openshift-monitoring/k8s",
+          "service": "dns-default"
+        },
+        "value": [
+          1705578106.559,
+          "1"
+        ]
+      },
+      {
+        "metric": {
+          "__name__": "up",
+          "container": "kube-rbac-proxy",
+          "endpoint": "metrics",
+          "instance": "10.128.2.5:9154",
+          "job": "dns-default",
+          "namespace": "openshift-dns",
+          "pod": "dns-default-mf6gk",
+          "prometheus": "openshift-monitoring/k8s",
+          "service": "dns-default"
+        },
+        "value": [
+          1705578106.559,
+          "1"
+        ]
+      },
+  ...
+```
+
+This metrics list all the instances that are scraped in the opesnhift-dns namespace and their ability to respond to the scrape request. 
+
+# Let's create a user defined metrics 
+
+First we create a workload that expose metrics 
+
+```
+cat <<EOF |oc create -f -
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: ns1
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  labels:
+    app: prometheus-example-app
+  name: prometheus-example-app
+  namespace: ns1
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: prometheus-example-app
+  template:
+    metadata:
+      labels:
+        app: prometheus-example-app
+    spec:
+      containers:
+      - image: ghcr.io/rhobs/prometheus-example-app:0.4.1
+        imagePullPolicy: IfNotPresent
+        name: prometheus-example-app
+---
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    app: prometheus-example-app
+  name: prometheus-example-app
+  namespace: ns1
+spec:
+  ports:
+  - port: 8080
+    protocol: TCP
+    targetPort: 8080
+    name: web
+  selector:
+    app: prometheus-example-app
+  type: ClusterIP
+EOF 
+```
+
+To check this pod create metrics we create a curl pod and grab the metrics using the service created previously:  
+```
+oc run curl --image=curlimages/curl --command --  tail -f /dev/null 
+oc rsh curl
+curl http://prometheus-example-app:8080/metrics
+```
+
+you should get this answer 
+```
+# HELP version Version information about this binary
+# TYPE version gauge
+version{version="v0.4.1"} 1
+```
+
+This metrics is of little interest but is simple to handle and understand
+
+# Add this metrics in Openshift Monitoring 
+
+## What is a ServiceMonitor
+
+A ServiceMonitor is a custom resource that participate to the configuration of Prometheus.
```

Comment:
```suggestion
A ServiceMonitor is a custom resource that feeds into the configuration of Prometheus.
```

Reply:
that composes part of a Prometheus configuration.
? is that a better way to phrase this?

Reply:
Yes I find `feeds into`slightly ambiguous

Reply:
I will opt for @mlavi suggestion